src/database-controller/src/core/framework.js

```diff
@@ -0,0 +1,421 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+const logger = require('@dbc/core/logger');
+const k8s = require('@dbc/core/k8s');
+const _ = require('lodash');
+const yaml = require('js-yaml');
+
+const mockFrameworkStatus = () => {
+  return {
+    state: 'AttemptCreationPending',
+    attemptStatus: {
+      completionStatus: null,
+      taskRoleStatuses: [],
+    },
+    retryPolicyStatus: {
+      retryDelaySec: null,
+      totalRetriedCount: 0,
+      accountableRetriedCount: 0,
+    },
+  };
+};
+
+const convertState = (state, exitCode, retryDelaySec) => {
+  switch (state) {
+    case 'AttemptCreationPending':
+    case 'AttemptCreationRequested':
+    case 'AttemptPreparing':
+      return 'WAITING';
+    case 'AttemptRunning':
+      return 'RUNNING';
+    case 'AttemptDeletionPending':
+    case 'AttemptDeletionRequested':
+    case 'AttemptDeleting':
+      if (exitCode === -210 || exitCode === -220) {
+        return 'STOPPING';
+      } else {
+        return 'RUNNING';
+      }
+    case 'AttemptCompleted':
+      if (retryDelaySec == null) {
+        return 'RUNNING';
+      } else {
+        return 'WAITING';
+      }
+    case 'Completed':
+      if (exitCode === 0) {
+        return 'SUCCEEDED';
+      } else if (exitCode === -210 || exitCode === -220) {
+        return 'STOPPED';
+      } else {
+        return 'FAILED';
+      }
+    default:
+      return 'UNKNOWN';
+  }
+};
+
+function ignoreError(err) {
+  logger.info('This error will be ignored: ', err);
+}
+
+// Class `Snapshot` handles the full json of framework.
+// It provides method like:
+//    getRequest: extract framework request from the full json
+//    overrideRequest: override the framework request using another snapshot
+//    getRequestUpdate, getStatusUpdate, getAllUpdate: Get database updates from the snapshot.
+// It doesn't handle database internal status, like: requestSynced, apiServerDeleted, ..., etc.
+class Snapshot {
+  constructor(snapshot) {
+    if (snapshot instanceof Object) {
+      this._snapshot = _.cloneDeep(snapshot);
+    } else {
+      this._snapshot = JSON.parse(snapshot);
+    }
+    // If the snapshot doesn't have a status, mock one instead.
+    // This usually happens when the framework spec is generated by rest-server.
+    if (!this._snapshot.status) {
+      this._snapshot.status = mockFrameworkStatus();
+    }
+  }
+
+  copy() {
+    return new Snapshot(this._snapshot);
+  }
+
+  getRequest(omitGeneration) {
+    const request = _.pick(this._snapshot, [
+      'apiVersion',
+      'kind',
+      'metadata.name',
+      'metadata.labels',
+      'metadata.annotations',
+      'spec',
+    ]);
+    if (omitGeneration) {
+      return _.omit(request, 'metadata.annotations.requestGeneration');
+    } else {
+      return request;
+    }
+  }
+
+  overrideRequest(otherSnapshot) {
```

Comment:
>overrideRequest [](start = 2, length = 15)

Can you leverage getRequest to do overrideRequest?
So no need to define request fields in 2 places

Reply:
I can't find a function to merge the request automatically in javascript, like: this._snapshot = MERGE(this.getRequest(), this._snapshot). So I write these fields twice.